graspy/embed/oosase.py

```diff
@@ -0,0 +1,416 @@
+        ids : array_like, length m
+            If self.semi_supervised the list must be of length m and contain the node ids for 
+            the nodes to embed. Otherwise, it is a no-op.
+
+        Returns
+        -------
+        oos_embedding : array, shape (m, d)
+            The embedding of the m out of sample vertices.
+        """
+
+        # Check if fit is already called
+        check_is_fitted(self, ["in_sample_embedding"], all_or_any=all)
+
+        n = self.in_sample_embedding.shape[0]
+
+        # Type checking
+        check_array(
+            X,
+            ensure_2d=False,
+            allow_nd=False,
+            ensure_min_samples=1,
+            ensure_min_features=n,
+        )
+
+        print(n)
+
+        if X.ndim is 1:
+            X = X.reshape((1, -1))
+            m = 1
+        elif X.shape[1] != n:
+            msg = "Similarity vector must be of length n"
+            raise ValueError(msg)
+        else:
+            m = X.shape[0]
+
+        row_sums = np.sum(X, axis=1)
+        if np.count_nonzero(row_sums) != m:
+            msg = (
+                "At least one adjacency vector is the zero vector."
+                + " It is recommended to first embed nodes with non-zero adjacency vectors"
+                + " with self.semi_supervised = True and then embed the nodes"
+                + " with zero adjacencies"
+            )
+            warnings.warn(msg, UserWarning)
+
+        oos_embedding = X @ np.linalg.pinv(self.in_sample_embedding).T
+
+        if self.semi_supervised:
+            if ids is None:
+                if set(self.in_sample_idx) == set(range(n)):
+                    ids = np.arange(n, n + m)
+                else:
+                    msg = "Semi supervised embedding without node ids."
+                    raise ValueError(msg)
+            elif len(ids) != m:
+                msg = "Length of node ids, {}, does not match number of out of sample nodes, {}".format(
+                    len(ids), m
+                )
+                raise ValueError(msg)
+
+            self.in_sample_embedding = np.concatenate(
+                (self.in_sample_embedding, oos_embedding), axis=0
+            )
+
+            self.id_to_index.update(dict(zip(range(n, n + m), ids)))
+            self.index_to_id.update(dict(zip(ids, range(n, n + m))))
+
+        return oos_embedding
+
+    def fit_predict(self, graph, edge_weight_attr=None):
+        """
+        Perform both in sample and out of sample adjacency spectral embedding. If
+        the graph is too big, this function will break.
+
+        Only unweighted graphs supported.
+
+        Parameters
+        ----------
+        graph : array-like or networkx.Graph
+            Input graph to embed.
+        edge_weight_attr : string
+            The edge weight attribute. If None, it is assumed the graph is binary.
+
+
+        Returns
+        -------
+        embedding : array
+            Embedding of n * m vertices.
```

Comment:
this confuses me

Reply:
as it should ;-)

should read `n_in_sample_verts + n_out_sample_verts`